From 33b09bd2938880fc90db754dadd9f121188427bb Mon Sep 17 00:00:00 2001
From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Tue, 11 Feb 2025 09:20:53 +0100
Subject: Revert "hostfs: fix string handling in __dentry_name()"

From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

This reverts commit 86ec56b25476758f708328b2eeed68918567efd0 which is
commit 60a6002432448bb3f291d80768ae98d62efc9c77 upstream.

It is reported to cause build issues and odds are the root problem isn't
really an issue on the 6.6.y branch anyway.  If it is, someone can
provide a working set of backported patches.

Reported-by: Guenter Roeck <linux@roeck-us.net>
Link: https://lore.kernel.org/r/cd10a924-ae65-4b02-aea2-e629947ca7a3@roeck-us.net
Cc: Hongbo Li <lihongbo22@huawei.com>
Cc: Christian Brauner <brauner@kernel.org>
Cc: Sasha Levin <sashal@kernel.org>
Cc: Al Viro <viro@zeniv.linux.org.uk>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---
 fs/hostfs/hostfs_kern.c |   27 +++++++++++++++++++++------
 1 file changed, 21 insertions(+), 6 deletions(-)

--- a/fs/hostfs/hostfs_kern.c
+++ b/fs/hostfs/hostfs_kern.c
@@ -93,17 +93,32 @@ __uml_setup("hostfs=", hostfs_args,
 static char *__dentry_name(struct dentry *dentry, char *name)
 {
 	char *p = dentry_path_raw(dentry, name, PATH_MAX);
-	struct hostfs_fs_info *fsi = dentry->d_sb->s_fs_info;
-	char *root = fsi->host_root_path;
-	size_t len = strlen(root);
+	char *root;
+	size_t len;
+	struct hostfs_fs_info *fsi;
 
-	if (IS_ERR(p) || len > p - name) {
+	fsi = dentry->d_sb->s_fs_info;
+	root = fsi->host_root_path;
+	len = strlen(root);
+	if (IS_ERR(p)) {
 		__putname(name);
 		return NULL;
 	}
 
-	memcpy(name, root, len);
-	memmove(name + len, p, name + PATH_MAX - p);
+	/*
+	 * This function relies on the fact that dentry_path_raw() will place
+	 * the path name at the end of the provided buffer.
+	 */
+	BUG_ON(p + strlen(p) + 1 != name + PATH_MAX);
+
+	strscpy(name, root, PATH_MAX);
+	if (len > p - name) {
+		__putname(name);
+		return NULL;
+	}
+
+	if (p > name + len)
+		strcpy(name + len, p);
 
 	return name;
 }
